### ISSUE TITLE
Missing HTTP header sanitization when reading values from `/etc/os-release`

### ISSUE THREAD

**Issue by nazarewk** [dotnet/runtime#102819]
### Describe the bug

TLDR;
1. a Linux distribution can use unicode characters as part of their release names and put them into `/etc/os-release` and/or `/etc/lsb-release`
2. some part of .NET Core is putting it as-is (without any sanitization whatsoever) into `User-Agent` HTTP header
3. then refusing to perform the request due to `Request headers must contain only ASCII characters.`.

At the same time users of the operating system can have no influence over it and no way to change it.

I suspect it is reading the full `PRETTY_NAME` string from `/etc/os-release` somewhere.

detailed description is available at https://github.com/microsoftgraph/msgraph-cli/issues/477

### To Reproduce

put "funny" characters into names in `/etc/os-release`:

```
> cat /etc/os-release
ANSI_COLOR="1;34"
BUG_REPORT_URL="[https://github.com/NixOS/nixpkgs/issues](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc)"
BUILD_ID="24.11.20240528.aae38d0"
DOCUMENTATION_URL="https://nixos.org/learn.html"
HOME_URL="https://nixos.org/"
ID=nixos
IMAGE_ID=""
IMAGE_VERSION=""
LOGO="nix-snowflake"
NAME=NixOS
PRETTY_NAME="NixOS 24.11 (Vicuña)"
SUPPORT_URL="https://nixos.org/community.html"
VERSION="24.11 (Vicuña)"
VERSION_CODENAME=vicuña
VERSION_ID="24.11"
```

### Further technical details

I have encountered it when using `msgraph-cli`, but seems like every .NET tool doing network requests is affected: [nuget](https://matrix.to/#/!RRerllqmbATpmbJgCn:nixos.org/$vvoGSryfYl9s-ApOXehZfKI5A57WN6rNQNpKa_xj950?via=nixos.org&via=matrix.org&via=tchncs.de) etc. in NixOS (a distribution mostly built from sources), 

I don't have any experience with .NET whatsoever, just filing a bug that I have encountered today.

**Comment by wfurt** [comment]
Do you have specific c# example @nazarewk? AFAIK neither `HttpClient` no `HttpWebRequest` populate `User-Agent` e.g. it feels the the problematic code does not live in runtime.

**Comment by rzikm** [comment]
> AFAIK neither HttpClient no HttpWebRequest populate User-Agent e.g. it feels the the problematic code does not live in runtime.

Correct, I just tested it with a HttpClient sample.

**Comment by nazarewk** [comment]
> Do you have specific c# example @nazarewk? AFAIK neither `HttpClient` no `HttpWebRequest` populate `User-Agent` e.g. it feels the the problematic code does not live in runtime.

I have absolutely no idea where it happens, https://github.com/microsoftgraph/msgraph-cli/issues/477#issuecomment-2137366803 pointed me to this repo.

**Comment by wfurt** [comment]
I understand that it may not be obvious where this is coming from. Just looking at the value "azsdk-net-Identity" it seems like some SDK that is bellow msgraph. I did search through https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/identity but I did not see any obvious hit. It _may_  come from some of the loaded plugins. You may try to put break on `SendAsync` and check the call stack. Or use something like "Fiddler" and look at URL this is trying to connect to. That may provide some insight about the offending code.

**Comment by nazarewk** [comment]
> I understand that it may not be obvious where this is coming from. Just looking at the value "azsdk-net-Identity" it seems like some SDK that is bellow msgraph. I did search through [Azure/azure-sdk-for-net@`main`/sdk/identity](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/identity?rgh-link-date=2024-05-29T17%3A17%3A57Z) but I did not see any obvious hit. It _may_ come from some of the loaded plugins. You may try to put break on `SendAsync` and check the call stack. Or use something like "Fiddler" and look at URL this is trying to connect to. That may provide some insight about the offending code.

I'm not a .NET programmer, I'm just compiling/packaging an existing tool so I can't really modify code to debug it. I know the request is to some access token issuance endpoint (it was during `mgc login`) and that the error also happens during downloading nuget modules (through somebody else reporting failure to build a .NET package on NixOS).

**Comment by wfurt** [comment]
Actually I think I found it

https://github.com/Azure/azure-sdk-for-net/blob/5004a5cb6202dab381d5a7a452da8ee775db3a0e/sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ClientLibraryInformation.cs#L56-L82

the `RuntimeInformation.OSDescription` makes no guarantee about HTTP safety -> "An opaque string that describes the operating system on which the app is running."

I cannot move this issue across project boundaries e.g. `dotnet` vs `Azure`.  I suggest you open yet another issue in https://github.com/Azure/azure-sdk-for-net/issues since you are original reporter @nazarewk and you can link it here and close this issue.

**Comment by nazarewk** [comment]
> Actually I think I found it
> 
> [Azure/azure-sdk-for-net@`5004a5c`/sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ClientLibraryInformation.cs#L56-L82](https://github.com/Azure/azure-sdk-for-net/blob/5004a5cb6202dab381d5a7a452da8ee775db3a0e/sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ClientLibraryInformation.cs#L56-L82)
> 
> the `RuntimeInformation.OSDescription` makes no guarantee about HTTP safety -> "An opaque string that describes the operating system on which the app is running."
> 
> I cannot move this issue across project boundaries e.g. `dotnet` vs `Azure`. I suggest you open yet another issue in [Azure/azure-sdk-for-net/issues](https://github.com/Azure/azure-sdk-for-net/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc) since you are original reporter @nazarewk and you can link it here and close this issue.

@wfurt thanks for the investigation, I have created https://github.com/Azure/azure-sdk-for-net/issues/44300 per your suggestion

**Comment by wfurt** [comment]
moved to Azure SDK

**Comment by purepani** [comment]
Hey @wfurt, we're also having trouble using the dotnet cli to install dotnet packages due to this error(which is fixed by changing the name to exclude non-ascii characters). I wouldn't expect that any of the dotnet tools would depend on the azure-sdk, so I wanted to ask about that.

**Comment by wfurt** [comment]
what error do you see @purepani? Since there was no change in runtime for this it should not matter. I'm not familia with Azure SDK release process so I cannot comment on that much. 

And did you verified this is the very same issue?

**Comment by purepani** [comment]
> what error do you see @purepani? Since there was no change in runtime for this it should not matter. I'm not familia with Azure SDK release process so I cannot comment on that much.
> 
> And did you verified this is the very same issue?

Hi,
These are the steps that we take, and what causes the error.
1. Use some distro with a non-ascii character in the PRETTY_NAME string(as listed in the original issue posting, the distro name for NixOS that causes the issue is `Vicuña`).
2. Find a dotnet project and use `dotnet restore`
3. Get the following error: `error NU1301: Unable to load the service index for source https://api.nuget.org/v3/index.json` for every package that attempts to get installed.

If we manually change the distro name to `Vicuna` in order to remove the non-ascii character, the issue resolves itself. 

I agree that the issue isn't necessarily in the runtime, so the issue may be better placed in a different dotnet project, but I don't think this specific problem is caused by the azure-sdk unless you depend on the azure-sdk to retrieve packages(which didn't seem like it particularly made sense to me). I think the core issue is the same about needing to sanitize headers to remove non-unicode characters.

**Comment by wfurt** [comment]
It seems like also `Nuget` puts the description there without any checks

https://github.com/NuGet/NuGet.Client/blob/dc2e3b27170c779e3c01a7306504fb83cb4ef147/src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/XPlatUtility.cs#L66

It should be fixed there. It is unfortunate IMHO but `HttpClient` is not responsible for creating the string.
